### PR TITLE
Update machine.Spec.ProviderID field

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
https://jira.coreos.com/browse/CLOUD-445

### CODE REVIEW

File: pkg/actuators/machine/actuator.go [hunk 1]
```diff
@@ -116,9 +117,39 @@ func (a *Actuator) Create(context context.Context, cluster *clusterv1.Cluster, m
 		}
 		return err
 	}
+	err = a.updateProviderID(machine, instance)
+	if err != nil {
+		return fmt.Errorf("failed to update machine object with providerID: %v", err)
+	}
 	return a.updateStatus(machine, instance)
```
Comment (frobware) [inline pkg/actuators/machine/actuator.go]: If we just updated the providerID isn't `machine` on this line a different version of the object now?

File: pkg/actuators/machine/actuator.go [hunk 2]
```diff
@@ -116,9 +117,39 @@ func (a *Actuator) Create(context context.Context, cluster *clusterv1.Cluster, m
...
+		}
+		machineCopy.Spec.ProviderID = &providerID
+		if err := a.client.Update(context.Background(), machineCopy); err != nil {
+			return fmt.Errorf("error updating machine spec ProviderID: %v", err)
+		}
+		glog.Infof("ProviderID updated at machine spec: %s", providerID)
+	} else {
+		machineCopy.Spec.ProviderID = nil
+		if err := a.client.Update(context.Background(), machineCopy); err != nil {
+			return fmt.Errorf("error updating machine annotations: %v", err)
```
Comment (frobware) [inline pkg/actuators/machine/actuator.go]: This message is a little misleading. s/annotations/ProviderID.

File: pkg/actuators/machine/actuator.go [hunk 3]
```diff
@@ -116,7 +117,37 @@ func (a *Actuator) Create(context context.Context, cluster *clusterv1.Cluster, m
...
+// updateProviderID adds providerID in the machine spec
+func (a *Actuator) updateProviderID(machine *machinev1.Machine, instance *ec2.Instance) (*machinev1.Machine, error) {
+	existingProviderID := machine.Spec.ProviderID
+	machineCopy := machine.DeepCopy()
+	if instance != nil {
+		providerID := fmt.Sprintf("aws:///%s/%s", aws.StringValue(instance.Placement.AvailabilityZone), aws.StringValue(instance.InstanceId))
+
+		if existingProviderID != nil && *existingProviderID == providerID {
+			glog.Infof("ProviderID already set in the machine Spec with value:%s", *existingProviderID)
+			return nil, nil
```
Comment (frobware) [inline pkg/actuators/machine/actuator.go]: Shouldn't this case return the original machine (as passed in)? We go on to use it unconditionally if there is no error:

```go
	updatedMachine, err := a.updateProviderID(machine, instance)
	if err != nil {
		return fmt.Errorf("failed to update machine object with providerID: %v", err)
	}
	return a.updateStatus(updatedMachine, instance)
```

**Comment by frobware** [comment]
What happens on upgrade?

- I have 4.1 with machine already running
- I take 4.2 as an upgrade

Will those existing machines be updated to now have provider id?

**Comment by vikaschoudhary16** [comment]
I have not tested upgrade scenario yet but i believe it should work because upgrade creates new machines. To be 100% sure, will test it out.

**Comment by frobware** [comment]
> I have not tested upgrade scenario yet but i believe it should work because upgrade creates new machines. To be 100% sure, will test it out.

I wasn't so concerned about new machines, it was existing machines that -- I believe -- would not get the provider id added.

**Comment by enxebre** [comment]
@vikaschoudhary16 @frobware We need to make sure consumers wanting to rely on providerID e.g autoscaler are backward compatible since old machines won't necessarily have it set

Also we probably want to force a migration path once the cvo fetch the new mao version during an upgrade - may be also ensuring provideID on `update()` so as the new controller comes up during a cluster upgrade it would reconcile them all

Also this PR depends on the the az field being set which we do in the installer but is not required
so we need to cover those cases
providerID might look as `aws:///zone/id` or `aws:////id` (edited) 
https://github.com/openshift/cluster-api-provider-aws/pull/210/files#diff-acdcd15b24a848e93a682408b101755fR132

**Comment by ingvagabund** [comment]
> Also this PR depends on the the az field being set which we do in the installer but is not required
so we need to cover those cases
providerID might look as aws:///zone/id or aws:////id (edited)

Pasting the line for better readability:

> providerID := fmt.Sprintf("aws:///%s/%s", aws.StringValue(instance.Placement.AvailabilityZone), aws.StringValue(instance.InstanceId))

It's already covered up to a case when `instance.Placement == nil`. In which case the code will just panic.

See https://github.com/openshift/cluster-api-provider-aws/pull/220

**Comment by enxebre** [comment]
placement is not a pointer, this won't panic. My concern is from a consumer pov

**Pull request merged**